pkg/utils/build/source_image.go

```diff
@@ -165,3 +176,274 @@ func IsPreFetchDependencysFilesExists(absExtraSourceDirPath string, isHermetic b
+	seen := make(map[string]int)
+	parentImages := d.ParentImages()
+	for _, imagePullspec := range parentImages {
+		if imagePullspec == "scratch" {
+			continue
+		}
+		if _, exists := seen[imagePullspec]; exists {
+			continue
+		}
+		seen[imagePullspec] = 1
+		if converted, err := convertImageToBuildahOutputForm(imagePullspec); err == nil {
+			convertedImagePullspecs = append(convertedImagePullspecs, converted)
+		} else {
+			return nil, err
+		}
+	}
+	return convertedImagePullspecs, nil
+}
+
+func isRegistryAllowed(registry string) bool {
+	// For the list of allowed registries, refer to source-build task definition.
+	allowedRegistries := map[string]int{
+		"registry.access.redhat.com": 1,
+		"registry.redhat.io":         1,
+	}
+	_, exists := allowedRegistries[registry]
+	return exists
+}
+
+func IsImagePulledFromAllowedRegistry(imagePullspec string) (bool, error) {
+	if ref, err := reference.Parse(imagePullspec); err == nil {
+		return isRegistryAllowed(ref.Registry), nil
+	} else {
+		return false, err
+	}
+}
+
+func SourceBuildTaskRunLogsContain(
+	tektonController *tekton.TektonController, pr *pipeline.PipelineRun, message string) (bool, error) {
+	logs, err := tektonController.GetTaskRunLogs(pr.GetName(), "build-source-image", pr.GetNamespace())
+	if err != nil {
+		return false, err
+	}
+	for _, logMessage := range logs {
+		if strings.Contains(logMessage, message) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func ResolveSourceImage(image string) (string, error) {
+	config, err := FetchImageConfig(image)
+	if err != nil {
+		return "", err
+	}
+	labels := config.Config.Labels
+	var version, release string
+	var exists bool
+	if version, exists = labels["version"]; !exists {
+		return "", fmt.Errorf("cannot find out version label from image config")
+	}
+	if release, exists = labels["release"]; !exists {
+		return "", fmt.Errorf("cannot find out release label from image config")
+	}
+	ref, err := reference.Parse(image)
+	if err != nil {
+		return "", err
+	}
+	ref.ID = ""
+	ref.Tag = fmt.Sprintf("%s-%s-source", version, release)
+	return ref.Exact(), nil
+}
+
+func AllParentSourcesIncluded(parentSourceImage, builtSourceImage string) (bool, error) {
+	parentConfig, err := FetchImageConfig(parentSourceImage)
+	if err != nil {
+		return false, err
+	}
+	builtConfig, err := FetchImageConfig(builtSourceImage)
+	if err != nil {
+		return false, err
+	}
+	srpmSha256Sums := make(map[string]int)
+	var parts []string
+	for _, history := range builtConfig.History {
+		// Example history: #(nop) bsi version 0.2.0-dev adding artifact: 5f526f4
+		parts = strings.Split(history.CreatedBy, " ")
+		// The last part 5f526f4 is the checksum calculated from the file included in the generated blob.
+		srpmSha256Sums[parts[len(parts)-1]] = 1
+	}
+	for _, history := range parentConfig.History {
+		parts = strings.Split(history.CreatedBy, " ")
+		if _, exists := srpmSha256Sums[parts[len(parts)-1]]; !exists {
+			return false, nil
```

Comment:
Didn't we have some issue where the checksums of the layers are different even though the SRPMs are the same? I guess that doesn't affect us in this case?

Reply:
This checksum is the one calculated from the SRPM itself, not the archive including SRPM and symlink.

Reply:
Ah, right :+1: